### PR TITLE
feat(compute): add compute samples

### PR DESCRIPTION
**Pull request by FrodoTheTrue**

### CODE REVIEW

File: compute/create_instance.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	}
+	defer instancesClient.Close()
+
+	req := &computepb.InsertInstanceRequest{
+		Project: projectID,
+		Zone:    zone,
+		InstanceResource: &computepb.Instance{
+			Name: proto.String(instanceName),
+			Disks: []*computepb.AttachedDisk{
+				&computepb.AttachedDisk{
```
Comment (codyoss) [inline compute/create_instance.go:51]: nit: `&computepb.AttachedDisk` can be removed

File: compute/create_instance.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+						SourceImage: proto.String(sourceImage),
+					},
+					AutoDelete: proto.Bool(true),
+					Boot:       proto.Bool(true),
+					Type:       computepb.AttachedDisk_PERSISTENT.Enum(),
+				},
+			},
+			MachineType: proto.String(fmt.Sprintf("zones/%s/machineTypes/%s", zone, machineType)),
+			NetworkInterfaces: []*computepb.NetworkInterface{
+				&computepb.NetworkInterface{
```
Comment (codyoss) [inline compute/create_instance.go:63]: nit: `&computepb.NetworkInterface` can be removed.

File: compute/create_instance.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+// [START compute_instances_create]
+import (
+	"context"
+	"fmt"
+	"io"
+
+	compute "cloud.google.com/go/compute/apiv1"
+	computepb "google.golang.org/genproto/googleapis/cloud/compute/v1"
+
```
Comment (codyoss) [inline compute/create_instance.go:25]: nit: remove extra space.

File: compute/create_instance.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+				&computepb.NetworkInterface{
+					Name: proto.String(networkName),
+				},
+			},
+		},
+	}
+
+	op, err := instancesClient.Insert(ctx, req)
+	if err != nil {
+		return fmt.Errorf("Create instances request: %v", err)
```
Comment (codyoss) [inline compute/create_instance.go:72]: Errors should start with a lowercase letter. Maybe something like `"unable to create instances: %v"`

[Error strings](https://github.com/golang/go/wiki/CodeReviewComments#error-strings)

Comment (codyoss) [reply compute/create_instance.go:72]: This I believe happens in all files, just leaving this one comment that applies to all of them.

File: compute/create_instance.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	if op.GetStatus() == computepb.Operation_RUNNING {
```
Comment (codyoss) [inline compute/create_instance.go:75]: Could this also be in pending? I wonder if it is safer to just always wait on the operation regardless of the status? Applies to all LRO operations.

File: compute/create_instance.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+		}
+		defer zoneOperationsClient.Close()
+
+		req := &computepb.WaitZoneOperationRequest{
+			Operation: op.GetName(),
+			Project:   projectID,
+			Zone:      zone,
+		}
+
+		zoneOperationsClient.Wait(ctx, req)
```
Comment (codyoss) [inline compute/create_instance.go:88]: This returns an operation and an error. The error is not being checked. This applies to other LRO operations as well.

File: compute/disable_usage_export.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+	defer projectsClient.Close()
+
+	// Updating the setting with empty UsageExportLocationResource will disable the usage report generation.
+	req := &computepb.SetUsageExportBucketProjectRequest{
+		Project:                     projectID,
+		UsageExportLocationResource: &computepb.UsageExportLocation{},
+	}
+
+	projectsClient.SetUsageExportBucket(ctx, req)
```
Comment (codyoss) [inline compute/disable_usage_export.go:44]: This error should be checked. Should we do something with the operation?

File: compute/get_usage_export_bucket.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+	"context"
+	"fmt"
+	"io"
+
+	compute "cloud.google.com/go/compute/apiv1"
+	computepb "google.golang.org/genproto/googleapis/cloud/compute/v1"
+	"google.golang.org/protobuf/proto"
+)
+
+// getUsageExportBucketrRetrieve Compute Engine usage export bucket for the Cloud project. Replaces the empty value returned by the API with the default value used to generate report file names.
```
Comment (codyoss) [inline compute/get_usage_export_bucket.go:28]: nit: typo

File: compute/list_all_instances.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+		return fmt.Errorf("NewInstancesRESTClient: %v", err)
+	}
+
+	defer projectsClient.Close()
+
+	req := &computepb.AggregatedListInstancesRequest{
+		Project: projectID,
+	}
+
+	resp, err := projectsClient.AggregatedList(ctx, req)
```
Comment (codyoss) [inline compute/list_all_instances.go:42]: We can include these for now, but I believe this method should be paginated. This is still being worked on right now for go. See https://github.com/googleapis/gapic-generator-go/pull/711. If we add listing samples they will need to be updated in not too long.

File: compute/list_instances.go [hunk 10]
```diff
@@ -29,19 +29,19 @@ func listInstances(w io.Writer, projectID string, zone string) error {
 	// projectID := "your_project_id"
 	// zone := "europe-central2-b"
 	ctx := context.Background()
-	client, err := compute.NewInstancesRESTClient(ctx)
+	projectsClient, err := compute.NewInstancesRESTClient(ctx)
```
Comment (codyoss) [inline compute/list_instances.go:32]: nit: This looks like an instance client actually.

File: compute/list_instances.go [hunk 11]
```diff
@@ -29,19 +29,19 @@ func listInstances(w io.Writer, projectID string, zone string) error {
...
 	if err != nil {
 		return fmt.Errorf("NewInstancesRESTClient: %v", err)
 	}
 
-	defer client.Close()
+	defer projectsClient.Close()
```
Comment (codyoss) [inline compute/list_instances.go:37]: nit: remove newline before this. Some other samples might need this as well. Should just be consistent either way.

File: compute/set_usage_export_bucket.go [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+			ReportNamePrefix: proto.String(reportNamePrefix),
+		},
+	}
+
+	if reportNamePrefix == "" {
+		// Sending an empty value for reportNamePrefix results in the next usage report being generated with the default prefix value "usage_gce". (see: https://cloud.google.com/compute/docs/reference/rest/v1/projects/get)
+		fmt.Fprintf(w, "Setting reportNamePrefix to empty value causes the report to have the default prefix value `usage_gce`.\n")
+	}
+
+	projectsClient.SetUsageExportBucket(ctx, req)
```
Comment (codyoss) [inline compute/set_usage_export_bucket.go:55]: Check the error at least, maybe use the operation?

File: compute/usage_export_test.go [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+	"testing"
+	"time"
+
+	compute "cloud.google.com/go/compute/apiv1"
+	"cloud.google.com/go/storage"
+	"github.com/GoogleCloudPlatform/golang-samples/internal/testutil"
+	computepb "google.golang.org/genproto/googleapis/cloud/compute/v1"
+)
+
+func assertEqual(t *testing.T, a interface{}, b interface{}) {
```
Comment (codyoss) [inline compute/usage_export_test.go:32]: Please inline this check.

Test failure formatting: https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures

File: compute/usage_export_test.go [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+	buf := &bytes.Buffer{}
+
+	if err := setUsageExportBucket(buf, tc.ProjectID, bucketName, ""); err != nil {
+		t.Errorf("setUsageExportBucket got err: %v", err)
+	}
+
+	expectedResult := "Setting reportNamePrefix to empty value causes the report to have the default prefix value `usage_gce`"
+	if got := buf.String(); !strings.Contains(got, expectedResult) {
+		t.Errorf("setUsageExportBucket got\n----\n%v\n----\nWant to contain:\n----\n%v\n----\n", got, expectedResult)
```
Comment (codyoss) [inline compute/usage_export_test.go:89]: Instead off all the extra formatting consider `%q` to quote the strings.

**Review by codyoss — CHANGES_REQUESTED** [review]
Thanks for the PR, a few comments.

File: compute/create_instance.go [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+)
+
+// createInstance sends an instance creation request to GCP and waits for it to complete.
+func createInstance(w io.Writer, projectID string, zone string, instanceName string, machineType string, sourceImage string, networkName string) error {
+	// projectID := "your_project_id"
+	// zone := "europe-central2-b"
+	// instanceName := "your_instance_name"
+	// machineType := "n1-standard-1"
+	// sourceImage := "projects/debian-cloud/global/images/family/debian-10"
+	// networkName := "global/networks/default"
```
Comment (Xmasreturns) [inline compute/create_instance.go:35]: Should we give some additional context here, like asking the reader to replace some values?

File: compute/delete_instance.go [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+import (
+	"context"
+	"fmt"
+	"io"
+
+	compute "cloud.google.com/go/compute/apiv1"
+	computepb "google.golang.org/genproto/googleapis/cloud/compute/v1"
+)
+
+// deleteInstance sends a delete request to GCP and waits for it to complete.
```
Comment (Xmasreturns) [inline compute/delete_instance.go:27]:
```suggestion
// deleteInstance sends a delete request to the Compute Engine API and waits for it to complete.
```

File: compute/disable_usage_export.go [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+func disableUsageExport(w io.Writer, projectID string) error {
+	// projectID := "your_project_id"
+	ctx := context.Background()
+	projectsClient, err := compute.NewProjectsRESTClient(ctx)
+	if err != nil {
+		return fmt.Errorf("NewProjectsRESTClient: %v", err)
+	}
+	defer projectsClient.Close()
+
+	// Updating the setting with empty UsageExportLocationResource will disable the usage report generation.
```
Comment (Xmasreturns) [inline compute/disable_usage_export.go:37]:
```suggestion
	// Providing an empty UsageExportLocationResource value disables the usage report generation.
```

File: compute/get_usage_export_bucket.go [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+	"context"
+	"fmt"
+	"io"
+
+	compute "cloud.google.com/go/compute/apiv1"
+	computepb "google.golang.org/genproto/googleapis/cloud/compute/v1"
+	"google.golang.org/protobuf/proto"
+)
+
+// getUsageExportBucket retrieve Compute Engine usage export bucket for the Cloud project. Replaces the empty value returned by the API with the default value used to generate report file names.
```
Comment (Xmasreturns) [inline compute/get_usage_export_bucket.go:28]:
```suggestion
// getUsageExportBucket retrieves the Compute Engine usage export bucket for the Cloud project. Replaces the empty value returned by the API with the default value used to generate report file names.
```

File: compute/get_usage_export_bucket.go [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+func getUsageExportBucket(w io.Writer, projectID string) error {
+	// projectID := "your_project_id"
+	ctx := context.Background()
+	projectsClient, err := compute.NewProjectsRESTClient(ctx)
+	if err != nil {
+		return fmt.Errorf("NewProjectsRESTClient: %v", err)
+	}
+	defer projectsClient.Close()
+
+	// Updating the setting with empty UsageExportLocationResource will disable the usage report generation.
```
Comment (Xmasreturns) [inline compute/get_usage_export_bucket.go:38]:
```suggestion
	// Updating the setting with an empty UsageExportLocationResource value disables the usage report generation.
```

File: compute/set_usage_export_bucket.go [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+	"context"
+	"fmt"
+	"io"
+
+	compute "cloud.google.com/go/compute/apiv1"
+	computepb "google.golang.org/genproto/googleapis/cloud/compute/v1"
+	"google.golang.org/protobuf/proto"
+)
+
+// setUsageExportBucket sets Compute Engine usage export bucket for the Cloud project. This sample presents how to interpret the default value for the report name prefix parameter.
```
Comment (Xmasreturns) [inline compute/set_usage_export_bucket.go:28]:
```suggestion
// setUsageExportBucket sets the Compute Engine usage export bucket for the Cloud project. This sample presents how to interpret the default value for the report name prefix parameter.
```

File: compute/set_usage_export_bucket.go [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+	// bucketName := "your_bucket_name"
+	// reportNamePrefix := ""
+	ctx := context.Background()
+	projectsClient, err := compute.NewProjectsRESTClient(ctx)
+	if err != nil {
+		return fmt.Errorf("NewProjectsRESTClient: %v", err)
+	}
+	defer projectsClient.Close()
+
+	// Updating the setting with empty UsageExportLocationResource will disable the usage report generation.
```
Comment (Xmasreturns) [inline compute/set_usage_export_bucket.go:40]:
```suggestion
	// Updating the setting with an empty UsageExportLocationResource value disables the usage report generation.
```

File: compute/list_all_instances.go [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+import (
+	"context"
+	"fmt"
+	"io"
+
+	compute "cloud.google.com/go/compute/apiv1"
+	computepb "google.golang.org/genproto/googleapis/cloud/compute/v1"
+)
+
+// listAllInstances gets all instances present in a project, grouped by their zone.
```
Comment (Xmasreturns) [inline compute/list_all_instances.go:27]: This just outputs a list of all instances, right? Using "gets all instances" here is a little ambiguous, considering "get" is separate functionality.

File: compute/create_instance.go [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+
+	ctx := context.Background()
+	instancesClient, err := compute.NewInstancesRESTClient(ctx)
+	if err != nil {
+		return fmt.Errorf("NewInstancesRESTClient: %v", err)
+	}
+	defer instancesClient.Close()
+
+	req := &computepb.InsertInstanceRequest{
```
Comment (Xmasreturns) [inline compute/create_instance.go:44]: Can you add a comment here that describes what's happening in this block?

In the create_instance sample, other language variations also mention that instance creation requires at least one persistent disk and one network interface

File: compute/create_instance.go [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
+		return fmt.Errorf("unable to create instance: %v", err)
+	}
+
+	zoneOperationsClient, err := compute.NewZoneOperationsRESTClient(ctx)
+	if err != nil {
+		return fmt.Errorf("NewZoneOperationsRESTClient: %v", err)
+	}
+	defer zoneOperationsClient.Close()
+
+	waitReq := &computepb.WaitZoneOperationRequest{
```
Comment (Xmasreturns) [inline compute/create_instance.go:80]: Does this need a comment to describe what it's waiting for or how long?

File: compute/create_instance.go [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+	"context"
+	"fmt"
+	"io"
+
+	compute "cloud.google.com/go/compute/apiv1"
+	computepb "google.golang.org/genproto/googleapis/cloud/compute/v1"
+	"google.golang.org/protobuf/proto"
+)
+
+// createInstance sends an instance creation request to GCP and waits for it to complete.
```
[the remark this answers is not in the document]
Comment (Xmasreturns) [reply compute/create_instance.go:28]: Also, it seems relevant to include that the instance gets created in the specified project and zone.

File: compute/create_instance.go (hunk 24 above)
Comment (FrodoTheTrue) [reply compute/create_instance.go]: Added

File: compute/create_instance.go (hunk 23 above)
Comment (FrodoTheTrue) [reply compute/create_instance.go:44]: Yep, here we are adding disk and interface. I guess request is simple.

File: compute/list_all_instances.go (hunk 22 above)
Comment (FrodoTheTrue) [reply compute/list_all_instances.go]: Replaced to "prints"

File: compute/create_instance.go (hunk 15 above)
Comment (codyoss) [reply compute/create_instance.go:35]: I think this is fine as is.

File: compute/disable_usage_export.go [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+	projectsClient, err := compute.NewProjectsRESTClient(ctx)
+	if err != nil {
+		return fmt.Errorf("NewProjectsRESTClient: %v", err)
+	}
+	defer projectsClient.Close()
+
+	// Providing an empty UsageExportLocationResource value disables the usage report generation.
+	req := &computepb.SetUsageExportBucketProjectRequest{
+		Project:                     projectID,
+		UsageExportLocationResource: &computepb.UsageExportLocation{},
```
Comment (codyoss) [inline compute/disable_usage_export.go:40]: If nothing is passed will this also disable the report? Or does an empty struct need to be sent?

File: compute/list_all_instances.go [hunk 27]
```diff
@@ -0,0 +1,61 @@
...
+
+	compute "cloud.google.com/go/compute/apiv1"
+	computepb "google.golang.org/genproto/googleapis/cloud/compute/v1"
+)
+
+// listAllInstances prints all instances present in a project, grouped by their zone.
+func listAllInstances(w io.Writer, projectID string) error {
+	// zone := "europe-central2-b"
+	ctx := context.Background()
+	projectsClient, err := compute.NewInstancesRESTClient(ctx)
```
Comment (codyoss) [inline compute/list_all_instances.go:31]: nit: InstanceClient

File: compute/set_usage_export_bucket.go [hunk 28]
```diff
@@ -0,0 +1,84 @@
...
+	"fmt"
+	"io"
+
+	compute "cloud.google.com/go/compute/apiv1"
+	computepb "google.golang.org/genproto/googleapis/cloud/compute/v1"
+	"google.golang.org/protobuf/proto"
+)
+
+// setUsageExportBucket sets the Compute Engine usage export bucket for the Cloud project. This sample presents how to interpret the default value for the report name prefix parameter.
+func setUsageExportBucket(w io.Writer, projectID string, bucketName string, reportNamePrefix string) error {
```
Comment (codyoss) [inline compute/set_usage_export_bucket.go:29]: nit: string only has to be stated the last time if projectID, bucketName, and reportNamePrefix are all strings.

File: compute/usage_export_test.go (hunk 13 above)
Comment (codyoss) [reply compute/usage_export_test.go:32]: Comment still applies.

File: compute/usage_export_test.go [hunk 29]
```diff
@@ -0,0 +1,155 @@
...
+	if a == b {
+		return
+	} else {
+		t.Errorf(fmt.Sprintf("Got: %d; want %d", a, b))
+	}
+}
+
+func createBucket(t *testing.T, projectID string, bucketName string) error {
+	ctx := context.Background()
+	tc := testutil.SystemTest(t)
```
Comment (codyoss) [inline compute/usage_export_test.go:42]: This is already called in the calling method, maybe pass it in if the projectID is needed.

File: compute/wait_for_operation.go [hunk 30]
```diff
@@ -0,0 +1,62 @@
...
+	"fmt"
+	"io"
+	"strings"
+
+	compute "cloud.google.com/go/compute/apiv1"
+	computepb "google.golang.org/genproto/googleapis/cloud/compute/v1"
+)
+
+// waitForOperation waits for an operation to be completed. Calling this function will block until the operation is finished.
+func waitForOperation(w io.Writer, op *computepb.Operation, projectID string) error {
```
Comment (codyoss) [inline compute/wait_for_operation.go:29]: nit: I think it would be better to pass the zone and name directly instead of an operation.

**Review by codyoss — CHANGES_REQUESTED** [review]
A couple more comments, but it looks good for the most part. Thanks!

File: compute/disable_usage_export.go (hunk 26 above)
Comment (FrodoTheTrue) [reply compute/disable_usage_export.go:40]: just tested it, it works without empty struct. (in Node.js for example it doesn't work, you need to pass empty object)

File: compute/usage_export_test.go [hunk 31]
```diff
@@ -107,8 +101,12 @@ func TestUsageExportSnippets(t *testing.T) {
 
 	usageExportLocation := project.GetUsageExportLocation()
 
-	assertEqual(t, *usageExportLocation.BucketName, bucketName)
-	assertEqual(t, *usageExportLocation.ReportNamePrefix, "")
+	if *usageExportLocation.BucketName != bucketName {
+		t.Errorf(fmt.Sprintf("Got: %s; want %s", *usageExportLocation.BucketName, bucketName))
+	}
+	if *usageExportLocation.ReportNamePrefix != "" {
+		t.Errorf(fmt.Sprintf("Got: %s; want %s", *usageExportLocation.BucketName, ""))
```
Comment (codyoss) [inline compute/usage_export_test.go:108]: nit:

```suggestion
		t.Errorf(fmt.Sprintf("Got: %s; want %q", *usageExportLocation.BucketName, ""))
```

**Pull request merged**